### PR TITLE
[orchestration] Add beta support for ods-infra type - master

### LINKED ISSUE #281: Support of infrastructure components next to application / test components
to support infrastructure components next to application / test components, additional features are required:

- new repo type "ods-infra"
- new TIP/TIR for infra
- adjust doc generator for ods-infra type (generate TIP/R - new templates!)
- adjust other docs to include qualified infrastructure reference for these components (BI-CS-AWS, BI-CS-AZURE, …)
- amend interface of ods-mro-jenkins-shared-lib and ods-jenkins-shared-lib 
 example(s):
-- https://github.com/opendevstack/ods-jenkins-shared-library/blob/master/vars/stageStartOpenshiftBuild.groovy#L42-L43 
-- https://github.com/opendevstack/ods-jenkins-shared-library/blob/master/src/org/ods/OdsContext.groovy#L638 
-- https://github.com/opendevstack/ods-mro-jenkins-shared-library/blob/master/src/org/ods/util/MROPipelineUtil.groovy#L90 

used then in TIR doc gen.. 
https://github.com/opendevstack/ods-mro-jenkins-shared-library/blob/master/src/org/ods/usecase/LeVADocumentUseCase.groovy#L935-L951

The big discussion will be who generates the docs (or better phrased, who triggers the doc gen .. the iaas quickstarter or the mro). This will impact massively what needs to change .. and how the interface changes

### PR DESCRIPTION
**Pull request by nichtraunzer**
Resolves #281

### CODE REVIEW

**Comment by nichtraunzer** [comment]
FYI @metmajer @clemensutschig @martsec

**Comment by clemensutschig** [comment]
@nichtraunzer  - cool, first PR! .. couple of comments
a) CodeNarc rule violations were found. See the report at: file:///home/runner/work/ods-jenkins-shared-library/ods-jenkins-shared-library/build/reports/codenarc/main.html - in the checks, you see the report attached,.. it's the whitespaces you introduced 

image

b) no changes in the leva document scheduler?

**Comment by nichtraunzer** [comment]
> @nichtraunzer - cool, first PR! .. couple of comments
> a) CodeNarc rule violations were found. See the report at: file:///home/runner/work/ods-jenkins-shared-library/ods-jenkins-shared-library/build/reports/codenarc/main.html - in the checks, you see the report attached,.. it's the whitespaces you introduced
> 
> b) no changes in the leva document scheduler?

add a) fixed

add b) babysteps ... need to be able to trigger the deployment of the AWS component first

File: src/org/ods/orchestration/util/MROPipelineUtil.groovy [hunk 1]
```diff
@@ -324,6 +325,15 @@ class MROPipelineUtil extends PipelineUtil {
                         } else {
                             this.logger.debug("Repo '${repo.id}' is of type ODS Code Component. Nothing to do in phase '${name}' for target environment '${targetEnvToken}'.")
                         }
+                    } else if (repo.type?.toLowerCase() == PipelineConfig.REPO_TYPE_ODS_INFRA) {
+                        this.steps.stage('ODS Infrastructure Component') {
+                            // FIXME: consult SMEs which STAGE we should use: BUILD or DEPLOY
```
Comment (metmajer) [inline src/org/ods/orchestration/util/MROPipelineUtil.groovy:330]: @michaelsauter asking for your support here. Currently, the blocks for every repo type look differently. Unsure about how the block for infrastructure components (e.g. Cloud services) should look like. Want to advise?

**Review by metmajer — CHANGES_REQUESTED** [review]
I have requested advice from @michaelsauter on how to best implement the execution of the new ODS_INFRA component type. Also, CodeNarc reports > 300 prio 3 issues in existing code. How to work around them?

File: src/org/ods/orchestration/util/MROPipelineUtil.groovy (hunk 1 above)
Comment (michaelsauter) [reply src/org/ods/orchestration/util/MROPipelineUtil.groovy:330]: I am not really sure here. In the end, it is up to the "repo type" to decide what to do and when to do it. It depends on what fits best to the type.

It looks like your are executing the `Jenkinsfile` here, which presumably deploys infrastructure? Then moving it to the `DEPLOY` step would sound logical to me.

Note that for the `ods` type, we build+deploy in `BUILD` when targeting DEV, and deploy in `DEPLOY` when targeting QA/PROD. This is not really great, and I'd love to split this when targeting DEV in the future (so that deployment also happens in `DEPLOY` for DEV).

File: build.gradle [hunk 2]
```diff
@@ -128,5 +128,5 @@ codenarcTest {
 }
 
 jacoco {
-  toolVersion = "0.8.3"
+  toolVersion = "0.8.6"
```
Comment (michaelsauter) [inline build.gradle:131]: I think it would be nice to keep these out of the PR here, and move it to a separate PR, as the changes look completely unrelated ...

**Comment by michaelsauter** [comment]
> Also, CodeNarc reports > 300 prio 3 issues in existing code. How to work around them?

Are these the existing issues, or are these new issues?

I think prio 3 is often debatable, and anyway we should tackle them in different PRs. We only prevent the build when prio 1 or 2 are violated, and for prio 3 we just ensure it doesn't get more with every change ...

File: src/org/ods/orchestration/util/MROPipelineUtil.groovy (hunk 1 above)
Comment (metmajer) [reply src/org/ods/orchestration/util/MROPipelineUtil.groovy:330]: In the end, we need to make sure that… the Jenkinsfile is executed for D/Q/P. Would you suggest to change the condition so that it is executed both in `build` and `deploy`? Or is `deploy` enough? How should this look like with the proper combination of `isAssembleMode` and `is PromotionMode`?

File: build.gradle (hunk 2 above)
Comment (metmajer) [reply build.gradle:131]: Sure, I had issues with Java 11 when using the old version on my local machine. Will put this into a different PR.

**Comment by metmajer** [comment]
@michaelsauter I assume this is based on the fact that the feature branch is quite old. Best approach IMHO would be to conclude our questions and last fixes here, and then create a new feature branch based on the latest state of `master` after the integration and release of ODS 3.1 features.

File: src/org/ods/orchestration/util/MROPipelineUtil.groovy (hunk 1 above)
Comment (michaelsauter) [reply src/org/ods/orchestration/util/MROPipelineUtil.groovy:330]: Only execute once. If you want to run for D/Q/P, then it does not matter which "mode" it is - meaning the condition should just consider if the phase and the repo type.

Comment (metmajer) [reply src/org/ods/orchestration/util/MROPipelineUtil.groovy:330]: @michaelsauter so you say that the code is good to stay as it is and where it is to be triggered when deploying to D/Q/P?

Comment (michaelsauter) [reply src/org/ods/orchestration/util/MROPipelineUtil.groovy:330]: @metmajer Yes to me it looks like this will be executed for D/Q/P and can stay like it is. You might want to execute it in the `DEPLOY` phase instead, but this is a matter of preference. I would tend to move it to `DEPLOY` but only my preference - both ways will work.

Comment (metmajer) [reply src/org/ods/orchestration/util/MROPipelineUtil.groovy:330]: @michaelsauter ah, now I got you. Yes, `DEPLOY` is where it should be (as you can see from my comment further above, I had assumed it was already). Cheers!!

Comment (metmajer) [reply src/org/ods/orchestration/util/MROPipelineUtil.groovy]: Resolved via https://github.com/opendevstack/ods-jenkins-shared-library/pull/388/commits/9cf62cedc9a7c1f9087348eeef5f5bdff0bcc024

File: build.gradle (hunk 2 above)
Comment (metmajer) [reply build.gradle]: Resolved via https://github.com/opendevstack/ods-jenkins-shared-library/pull/388/commits/b1f7ec500e6e6f3d41bcd91be7f3d9140c922262

**Comment by metmajer** [comment]
Nice work, @nichtraunzer, Josef Hartmann and Frank Joas on the ODS integration of a new infrastructure component type 👍 Now, we'll wait for ODS 3.1 to be released and merged mainline. Then, we'll update this branch from mainline. Until then, please keep testing your work with this branch. Thanks!

File: src/org/ods/orchestration/util/MROPipelineUtil.groovy [hunk 3]
```diff
@@ -324,6 +325,14 @@ class MROPipelineUtil extends PipelineUtil {
                         } else {
                             this.logger.debug("Repo '${repo.id}' is of type ODS Code Component. Nothing to do in phase '${name}' for target environment '${targetEnvToken}'.")
                         }
+                    } else if (repo.type?.toLowerCase() == PipelineConfig.REPO_TYPE_ODS_INFRA) {
+                        this.steps.stage('ODS Infrastructure Component') {
```
Comment (michaelsauter) [inline src/org/ods/orchestration/util/MROPipelineUtil.groovy:329]: The `stage` block should be removed. There are 6 stages: init, build, deploy, test, release, finalise. Having this `stage` block here would wrap the `ODS Infrastructure Component` inside the `deploy` stage (opened in `DeployStage`). I suspect this is a leftover from an old ODS version, which had substages for other repo types as well, which is now fixed.

File: src/org/ods/orchestration/Stage.groovy [hunk 4]
```diff
@@ -96,16 +98,21 @@ class Stage {
         def testReportsPath = "${PipelineUtil.XUNIT_DOCUMENTS_BASE_DIR}/${repo.id}/${type}"
 
         logger.debug("Collecting JUnit XML Reports ('${type}') for ${repo.id}")
+        
         def testReportsStashName = "test-reports-junit-xml-${repo.id}-${steps.env.BUILD_ID}"
         if (type != 'unit') {
             testReportsStashName = "${type}-${testReportsStashName}"
         }
+
+        logger.info("XXX logInfo '${repo.id} / ${type} / ${testReportsStashName}'")
```
Comment (clemensutschig) [inline src/org/ods/orchestration/Stage.groovy:107]: remove ...

File: src/org/ods/orchestration/usecase/DocGenUseCase.groovy [hunk 5]
```diff
@@ -57,51 +57,56 @@ abstract class DocGenUseCase {
...
     @SuppressWarnings(['JavaIoPackageAccess'])
-    String createOverallDocument(String coverType, String documentType, Map metadata,Closure visitor = null, String watermarkText = null) {
+    String createOverallDocument(String templateName, String documentType, Map metadata, Closure visitor = null, String watermarkText = null) {
         def documents = []
         def sections = []
 
         this.project.repositories.each { repo ->
-            def documentName = repo.data.documents[documentType]
 
+            this.steps.echo "XXX createOverallDocument - Repo.data       ${repo.data} "
```
Comment (clemensutschig) [inline src/org/ods/orchestration/usecase/DocGenUseCase.groovy:103]: remove

File: src/org/ods/orchestration/util/Project.groovy [hunk 6]
```diff
@@ -168,9 +168,17 @@ class Project {
         @NonCPS
         // FIXME: why can we not invoke derived methods in short form, e.g. .resolvedBugs?
         private List<JiraDataItem> getResolvedReferences(String type) {
-            // Reference this within jiraResolved (contains readily resolved references to other entities)
+            Project.this.steps.echo("XXX ")
```
Comment (clemensutschig) [inline src/org/ods/orchestration/util/Project.groovy:171]: remove ...

**Review by clemensutschig — CHANGES_REQUESTED** [review]

**Comment by clemensutschig** [comment]
and @metmajer codenArc also failed

File: src/org/ods/orchestration/util/Project.groovy (hunk 6 above)
Comment (metmajer) [reply src/org/ods/orchestration/util/Project.groovy]: Resolved via https://github.com/opendevstack/ods-jenkins-shared-library/pull/388/commits/74b60b1987e56618be81b04eed73690e6392c2a7

File: src/org/ods/orchestration/usecase/DocGenUseCase.groovy (hunk 5 above)
Comment (metmajer) [reply src/org/ods/orchestration/usecase/DocGenUseCase.groovy]: Resolved via https://github.com/opendevstack/ods-jenkins-shared-library/pull/388/commits/74b60b1987e56618be81b04eed73690e6392c2a7

File: src/org/ods/orchestration/Stage.groovy (hunk 4 above)
Comment (metmajer) [reply src/org/ods/orchestration/Stage.groovy]: Resolved via https://github.com/opendevstack/ods-jenkins-shared-library/pull/388/commits/74b60b1987e56618be81b04eed73690e6392c2a7

File: src/org/ods/orchestration/util/MROPipelineUtil.groovy (hunk 3 above)
Comment (metmajer) [reply src/org/ods/orchestration/util/MROPipelineUtil.groovy]: Resolved via aeec7628b3797aa33e7e369f884efe0c1a4c7203

**Comment by metmajer** [comment]
@clemensutschig hope aeec7628b3797aa33e7e369f884efe0c1a4c7203 resolves handling of build/deploy. Thoughts?

**Comment by metmajer** [comment]
@clemensutschig @michaelsauter we are still getting a negative CodeNarc report although all Class 1 and 2 issues have been fixed. Class 3 not. Thoughts?

**Comment by michaelsauter** [comment]
@metmajer The `build.gradle`file in your last commit (https://github.com/opendevstack/ods-jenkins-shared-library/blob/ae0d1d4d16c38c014caff35c642422082090a362/build.gradle#L122) declares a maximum of 321 violations of priority 3, but the report (https://github.com/opendevstack/ods-jenkins-shared-library/suites/1926270147/artifacts/37954716) shows 327. Either fix the introduced violations or bump the number in `build.gradle` :)

**Comment by clemensutschig** [comment]
> @clemensutschig hope [aeec762](https://github.com/opendevstack/ods-jenkins-shared-library/commit/aeec7628b3797aa33e7e369f884efe0c1a4c7203) resolves handling of build/deploy. Thoughts?

that looks good to me 🗡️

**Comment by clemensutschig** [comment]
```
library/test/groovy/org/ods/orchestration/scheduler/DocGenSchedulerSpec.groovy: 18: Can't have an abstract method in a non-abstract class. The class 'org.ods.orchestration.scheduler.DocGenSchedulerSpec$DocGenUseCaseImpl' must be declared abstract or the method 'boolean shouldCreateArtifact(java.lang.String, java.util.Map)' must be implemented.
> Task :compileTestGroovy FAILED
 @ line 18, column 5.
       class DocGenUseCaseImpl extends DocGenUseCase {
       ^

1 error
```

**Comment by metmajer** [comment]
Looking into it

**Comment by metmajer** [comment]
@nichtraunzer ready to try Q/P deployments!

**Comment by clemensutschig** [comment]
@metmajer  coool! - can you folks (a) test it and (b) also bump the number of prio3 findings now down to what it is .. ?

File: src/org/ods/orchestration/util/MROPipelineUtil.groovy [hunk 7]
```diff
@@ -324,6 +326,14 @@ class MROPipelineUtil extends PipelineUtil {
                         } else {
                             this.logger.debug("Repo '${repo.id}' is of type ODS Code Component. Nothing to do in phase '${name}' for target environment '${targetEnvToken}'.")
                         }
+                    } else if (repo.type?.toLowerCase() == PipelineConfig.REPO_TYPE_ODS_INFRA) {
+                        if (this.project.isAssembleMode && name == PipelinePhases.BUILD) {
+                            executeODSComponent(repo, baseDir)
+                        } else if (this.project.isPromotionMode && name == PipelinePhases.BUILD) {
```
Comment (metmajer) [inline src/org/ods/orchestration/util/MROPipelineUtil.groovy:332]: @clemensutschig `ods` repo types deploy to Q/P in `Phase.DEPLOY`. Won't a promotion in `Phase.BUILD` rule out orchestration of code and infra components?

**Review by clemensutschig — APPROVED** [review]

**Comment by metmajer** [comment]
@nichtraunzer @frankjoas I am closing this issue, since the code has been merged into 3.x through #570
